File: Tutorials/VB/Tutorials/Tutorials/Tutorial06.vb
Imports NetOffice
Imports Excel = NetOffice.ExcelApi

Public Class Tutorial06
    Implements ITutorial

    Dim _hostApplication As IHost

#Region "ITutorial Member"

    Public Sub Run() Implements TutorialsBase.ITutorial.Run

        ' this examples shows how i can use variant types(object in NetOffice) at runtime
        ' the reason for the most variant definitions in office is a more flexible value set.(95%)
        ' here is the code to demonstrate this

        ' start application
        Dim application As New Excel.Application()
        application.DisplayAlerts = False

        ' create new Workbook and a new named style
        Dim book As Excel.Workbook = application.Workbooks.Add()
        Dim sheet As Excel.Worksheet = book.Worksheets(1)
        Dim range As Excel.Range = sheet.Cells(1, 1)
        Dim myStyle As Excel.Style = book.Styles.Add("myUniqueStyle")

        ' Range.Style is defined as Variant in Excel and represents as object in NetOffice
        ' You got always an Excel.Style instance if you ask for
        Dim style As Excel.Style = range.Style

        'and here comes the magic. both sets are valid because the variants was very flexible in the setter
        range.Style = "myUniqueStyle"
        range.Style = myStyle

        ' Name, Bold, Size are string, bool and double but defined as Variant 
        style.Font.Name = "Arial"
        style.Font.Bold = True ' you can also set "True" and it works. variants makes it possible
        style.Font.Size = 14

        ' quit & dipose
        application.Quit()
        application.Dispose()

        _hostApplication.ShowFinishDialog()

    End Sub

    Public ReadOnly Property Caption As String Implements TutorialsBase.ITutorial.Caption
        Get
            Return "Tutorial06"
        End Get
    End Property

    Public ReadOnly Property Description As String Implements TutorialsBase.ITutorial.Description
        Get
            Return IIf(_hostApplication.LCID = 1033, "Understanding Variant", "Verstehen und verwenden von Variant Typen")
        End Get
    End Property

    Public Sub Connect(ByVal hostApplication As TutorialsBase.IHost) Implements TutorialsBase.ITutorial.Connect

        _hostApplication = hostApplication

    End Sub

    Public Sub ChangeLanguage(ByVal lcid As Integer) Implements TutorialsBase.ITutorial.ChangeLanguage

    End Sub

    Public Sub Disconnect() Implements TutorialsBase.ITutorial.Disconnect

    End Sub

    Public ReadOnly Property Panel As System.Windows.Forms.UserControl Implements TutorialsBase.ITutorial.Panel
        Get
            Return Nothing
        End Get
    End Property


    Public ReadOnly Property Uri As String Implements TutorialsBase.ITutorial.Uri
        Get
            Return IIf(_hostApplication.LCID = 1033, "http://netoffice.codeplex.com/wikipage?title=Tutorial06_EN_VB", "http://netoffice.codeplex.com/wikipage?title=Tutorial06_DE_VB")
        End Get
    End Property

#End Region

End Class

File: Tutorials/C#/Tutorials/Tutorials/Tutorial06.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using TutorialsBase;

using NetOffice;
using Excel = NetOffice.ExcelApi;

namespace TutorialsCS4
{
    public partial class Tutorial06 : ITutorial 
    {
        IHost _hostApplication;

        #region ITutorial Member

        public void Run()
        {
            // start application
            Excel.Application application = new Excel.Application();
            application.DisplayAlerts = false;

            // create new Workbook
            Excel.Workbook book = application.Workbooks.Add();
            Excel.Worksheet sheet = (Excel.Worksheet)book.Worksheets[1];
            Excel.Range range = sheet.Cells[1, 1];

            // Style is defined as Variant in Excel and represents as object in NetOffice
            // You can cast them at runtime without problems
            Excel.Style style = (Excel.Style)range.Style;

            // variant types can be a scalar type at runtime
            // another example way to use is 
            if (range.Style is string)
            {
                string myStyle = range.Style as string;
            }
            else if (range.Style is Excel.Style)
            {
                Excel.Style myStyle = (Excel.Style)range.Style;
            }

            // Name, Bold, Size are bool but defined as Variant and also converted to object
            style.Font.Name = "Arial";
            style.Font.Bold = true;
            style.Font.Size = 14;


            // Please note: the reason for the most variant definition is a more flexible value set.
            // the Style property from Range returns always a Style object
            // but if you have a new named style created with the name "myStyle" you can set range.Style = myNewStyleObject; or range.Style = "myStyle"
            // this kind of flexibility is the primary reason for Variants in Office
            // in any case, you dont lost the COM Proxy management from NetOffice for Variants. 

            // quit & dipose
            application.Quit();
            application.Dispose();

            _hostApplication.ShowFinishDialog();
        }

        public void Connect(IHost hostApplication)
        {
            _hostApplication = hostApplication;
        }

        public void Disconnect()
        {

        }

        public void ChangeLanguage(int lcid)
        {

        }

        public string Uri
        {
            get { return _hostApplication.LCID == 1033 ? "http://netoffice.codeplex.com/wikipage?title=Tutorial06_EN_CS" : "http://netoffice.codeplex.com/wikipage?title=Tutorial06_DE_CS"; }
        }

        public string Caption
        {
            get { return "Tutorial06"; }
        }

        public string Description
        {
            get { return _hostApplication.LCID == 1033 ? "Understanding Variant" : "Verstehen und verwenden von Variant Typen"; }
        }

        public UserControl Panel
        {
            get { return null; }
        }

        #endregion
    }
}
